src/core/AudioEngine.cpp

```diff
@@ -250,31 +298,128 @@ bool AudioEngine::criticalXRuns() const
+        return;
+    }
+    
+    if (availableSpace < _frames)
+	{
+        std::size_t framesToWrite = availableSpace;
+        std::size_t framesWritten = m_inputAudioRingBuffer->write(_ab, framesToWrite, false);
+        fprintf(stderr, "AudioEngine: Partial input ring buffer overflow, %zu of %zu frames written, %zu frames dropped.\n", framesWritten, _frames, _frames - framesWritten);
+    } else
+	{
+        std::size_t framesWritten = m_inputAudioRingBuffer->write(_ab, _frames, false);
+        if (framesWritten < _frames)
+		{
+            fprintf(stderr, "AudioEngine: Unexpected ring buffer write failure, %zu of %zu frames written.\n", framesWritten, _frames);
+        }
+    }
+}
+
+
+
+void AudioEngine::processBufferedInputFrames()
+{
+	if (!m_inputAudioRingBufferReader || !m_inputAudioRingBuffer) return;
+
+	std::size_t availableInRing = m_inputAudioRingBufferReader->read_space();
+	if (availableInRing == 0)
+	{
+		return;
+	}
+
+	const std::size_t maxFramesPerCall = DEFAULT_BUFFER_SIZE * 2;
+    std::size_t framesToProcess = std::min(availableInRing, maxFramesPerCall);
+
+	ringbuffer_reader_t<SampleFrame>::read_sequence_t sequence = m_inputAudioRingBufferReader->read_max(framesToProcess);
+
+	std::size_t framesInSequence = sequence.size();
+	if (framesInSequence == 0) return;
+
+	if (m_tempInputProcessingBuffer.capacity() < framesInSequence)
+	{
+        m_tempInputProcessingBuffer.reserve(framesInSequence * 2);
+    }
+    m_tempInputProcessingBuffer.resize(framesInSequence);
+
+	std::size_t len1 = sequence.first_half_size();
+    std::size_t len2 = sequence.second_half_size();
+
+	if (len2 == 0)
+	{
+        memcpy(m_tempInputProcessingBuffer.data(), sequence.first_half_ptr(), len1 * sizeof(SampleFrame));
+    } else
+	{
+        memcpy(m_tempInputProcessingBuffer.data(), sequence.first_half_ptr(), len1 * sizeof(SampleFrame));
+        memcpy(m_tempInputProcessingBuffer.data() + len1, sequence.second_half_ptr(), len2 * sizeof(SampleFrame));
+    }
+
 	requestChangeInModel();
 
-	f_cnt_t frames = m_inputBufferFrames[ m_inputBufferWrite ];
-	auto size = m_inputBufferSize[m_inputBufferWrite];
-	SampleFrame* buf = m_inputBuffer[ m_inputBufferWrite ];
+	SampleFrame* currentWriteBufPtr = m_inputBuffer[m_inputBufferWrite];
+    f_cnt_t currentWriteBufSize = m_inputBufferSize[m_inputBufferWrite];
+    f_cnt_t currentFramesInWriteBuf = m_inputBufferFrames[m_inputBufferWrite];
 
-	if( frames + _frames > size )
+	f_cnt_t requiredSize = currentFramesInWriteBuf + static_cast<f_cnt_t>(framesInSequence);
+
+	if (requiredSize > currentWriteBufSize)
 	{
-		size = std::max(size * 2, frames + _frames);
-		auto ab = new SampleFrame[size];
-		memcpy( ab, buf, frames * sizeof( SampleFrame ) );
-		delete [] buf;
+        f_cnt_t newSize = std::max(currentWriteBufSize * 2, requiredSize + DEFAULT_BUFFER_SIZE);
+        
+        auto newBuffer = new SampleFrame[newSize];
```

Comment:
If this function is run in realtime context, then this line would cause an allocation, which is not realtime-safe.